File: src/main/java/miragefairy2019/mod/modules/fairy/AbilitySet.java
package miragefairy2019.mod.modules.fairy;

import miragefairy2019.mod.api.fairy.IAbilitySet;
import miragefairy2019.mod.api.fairy.IAbilityType;
import mirrg.boron.util.struct.ImmutableArray;
import mirrg.boron.util.struct.Tuple;
import mirrg.boron.util.suppliterator.ISuppliterator;

import java.util.HashMap;
import java.util.Map;

public final class AbilitySet implements IAbilitySet {

    private final ImmutableArray<Tuple<IAbilityType, Double>> tuples;
    private Map<IAbilityType, Double> map = new HashMap<>();

    public AbilitySet(ImmutableArray<Tuple<IAbilityType, Double>> tuples) {
        this.tuples = tuples;
        for (Tuple<IAbilityType, Double> tuple : tuples) {
            map.put(tuple.x, tuple.y);
        }
    }

    @Override
    public ISuppliterator<Tuple<IAbilityType, Double>> getAbilities() {
        return tuples.suppliterator();
    }

    @Override
    public double getAbilityPower(IAbilityType abilityType) {
        return map.getOrDefault(abilityType, 0.0);
    }

}

File: src/main/java/miragefairy2019/modkt/impl/fairy/AbilitySet.kt
package miragefairy2019.modkt.impl.fairy

import miragefairy2019.modkt.api.fairy.IAbilityEntry
import miragefairy2019.modkt.api.fairy.IAbilitySet
import miragefairy2019.modkt.api.fairy.IAbilityType

class AbilitySet(private val list: Iterable<IAbilityEntry>) : IAbilitySet {
    private val map = list.associateBy { it.type }
    override fun getEntries() = list
    override fun getPower(type: IAbilityType) = map[type]?.power ?: 0.0
}

class AbilityEntry(private val type: IAbilityType, private val power: Double) : IAbilityEntry {
    override fun getPower() = power
    override fun getType() = type
}
